src/wavm/WAVMWasmModule.cpp

```diff
@@ -1494,4 +1592,89 @@ std::unique_ptr<openmp::PlatformThreadPool>& WAVMWasmModule::getOMPPool()
+{
+    printf("\n------ Module debug info ------\n");
+
+    if (isBound()) {
+        size_t memSizeBytes = defaultMemory->numPages * WASM_BYTES_PER_PAGE;
+
+        I32 stackPointer =
+          getModuleStackPointer(moduleInstance, executionContext);
+
+        I32 heapBase = getGlobalI32("__heap_base", executionContext);
+        I32 dataEnd = getGlobalI32("__data_end", executionContext);
+
+        size_t heapSizeBytes = memSizeBytes - heapBase;
+        size_t stackSizeBytes = stackPointer;
+        size_t dataSizeBytes = dataEnd - stackPointer;
+
+        float memSizeMb = ((float)memSizeBytes) / (1024 * 1024);
+        float heapSizeMb = ((float)heapSizeBytes) / (1024 * 1024);
+        float stackSizeMb = ((float)stackSizeBytes) / (1024 * 1024);
+        float dataSizeMb = ((float)dataSizeBytes) / (1024 * 1024);
+
+        Uptr tableSize = Runtime::getTableNumElements(defaultTable);
+
+        printf("Bound user:         %s\n", boundUser.c_str());
+        printf("Bound function:     %s\n", boundFunction.c_str());
+        printf("Stack pointer:      %i\n", stackPointer);
+        printf("Total memory:       %.3f MiB (%lu bytes)\n",
+               memSizeMb,
+               memSizeBytes);
+        printf("Stack size:         %.3f MiB (%lu bytes)\n",
+               stackSizeMb,
+               stackSizeBytes);
+        printf("Data size:          %.3f MiB (%lu bytes)\n",
+               dataSizeMb,
+               dataSizeBytes);
+        printf("Heap size:          %.3f MiB (%lu bytes)\n",
+               heapSizeMb,
+               heapSizeBytes);
+        printf("Stack range:        %i - %i\n", 0, stackPointer);
+        printf("Data range:         %i - %lu\n",
+               stackPointer,
+               stackPointer + dataSizeBytes);
+        printf("Heap range:         %i - %lu\n", heapBase, memSizeBytes);
+        printf("Table size:         %lu\n", tableSize);
+        printf("Dynamic modules:    %lu\n", dynamicModuleMap.size());
+
+        for (auto p : dynamicModuleMap) {
+            p.second.printDebugInfo(executionContext);
+        }
+
+        filesystem.printDebugInfo();
+    } else {
+        printf("Unbound\n");
+    }
+
+    printf("-------------------------------\n");
+
+    fflush(stdout);
+}
+
+uint32_t WAVMWasmModule::createMemoryGuardRegion()
+{
+    auto logger = faabric::util::getLogger();
+
+    uint32_t regionSize = 10 * faabric::util::HOST_PAGE_SIZE;
+
+    uint32_t wasmOffset = mmapMemory(regionSize);
+
+    uint8_t* nativePtr =
+      &Runtime::memoryRef<uint8_t>(defaultMemory, wasmOffset);
+
+    int res = mprotect(nativePtr, regionSize, PROT_NONE);
+    if (res != 0) {
+        logger->error("Failed to create memory guard: {}",
+                      std::strerror(errno));
+        throw std::runtime_error("Failed to create memory guard");
+    }
+
+    logger->debug(
+      "Created guard region {}-{}", wasmOffset, wasmOffset + regionSize);
+
+    return wasmOffset;
```

Comment:
Difficult to test that these guard regions actually work, as trying to write to them causes a `SIGSEGV`